File: src/dotnet/PloyBus.Tests/Transport/Transactions/Messages/Handlers/Error/ErrorHandlerTests.cs
using NUnit.Framework;

namespace PolyBus.Transport.Transactions.Messages.Handlers.Error;

[TestFixture]
public class ErrorHandlerTests
{
    private TestBus _testBus = null!;
    private IncomingMessage _incomingMessage = null!;
    private IncomingTransaction _transaction = null!;
    private TestableErrorHandler _errorHandler = null!;

    [SetUp]
    public void SetUp()
    {
        _testBus = new TestBus("TestBus");
        _testBus.Messages.Add(typeof(ErrorHandlerTestMessage));
        _incomingMessage = new IncomingMessage(
            bus: _testBus,
            body: "{}",
            messageInfo: _testBus.Messages.GetMessageInfo(typeof(ErrorHandlerTestMessage)));
        _transaction = new IncomingTransaction(_testBus, _incomingMessage);
        _errorHandler = new TestableErrorHandler();
    }

    [Test]
    public async Task Retrier_SucceedsOnFirstAttempt_DoesNotRetry()
    {
        // Arrange
        var nextCalled = false;

        Task Next()
        {
            nextCalled = true;
            return Task.CompletedTask;
        }

        // Act
        await _errorHandler.Retrier(_transaction, Next);

        // Assert
        Assert.That(nextCalled, Is.True);
        Assert.That(_transaction.OutgoingMessages.Count, Is.EqualTo(0));
    }

    [Test]
    public async Task Retrier_FailsOnce_RetriesImmediately()
    {
        // Arrange
        var callCount = 0;

        Task Next()
        {
            callCount++;
            if (callCount == 1)
            {
                throw new Exception("Test error");
            }

            return Task.CompletedTask;
        }

        // Act
        await _errorHandler.Retrier(_transaction, Next);

        // Assert
        Assert.That(callCount, Is.EqualTo(2));
        Assert.That(_transaction.OutgoingMessages.Count, Is.EqualTo(0));
    }

    [Test]
    public async Task Retrier_FailsAllImmediateRetries_SchedulesDelayedRetry()
    {
        // Arrange
        var expectedRetryTime = DateTime.UtcNow.AddMinutes(5);
        _errorHandler.SetNextRetryTime(expectedRetryTime);

        var callCount = 0;

        Task Next()
        {
            callCount++;
            throw new Exception("Test error");
        }

        // Act
        await _errorHandler.Retrier(_transaction, Next);

        // Assert
        Assert.That(callCount, Is.EqualTo(_errorHandler.ImmediateRetryCount));
        Assert.That(_transaction.OutgoingMessages.Count, Is.EqualTo(1));

        var delayedMessage = _transaction.OutgoingMessages[0];
        Assert.That(delayedMessage.DeliverAt, Is.EqualTo(expectedRetryTime));
        Assert.That(delayedMessage.Headers[_errorHandler.RetryCountHeader], Is.EqualTo("1"));
        Assert.That(delayedMessage.Endpoint, Is.EqualTo("TestBus"));
    }

    [Test]
    public async Task Retrier_WithExistingRetryCount_IncrementsCorrectly()
    {
        // Arrange
        _incomingMessage.Headers[_errorHandler.RetryCountHeader] = "2";
        var expectedRetryTime = DateTime.UtcNow.AddMinutes(10);
        _errorHandler.SetNextRetryTime(expectedRetryTime);

        Task Next() => throw new Exception("Test error");

        // Act
        await _errorHandler.Retrier(_transaction, Next);

        // Assert
        Assert.That(_transaction.OutgoingMessages.Count, Is.EqualTo(1));

        var delayedMessage = _transaction.OutgoingMessages[0];
        Assert.That(delayedMessage.Headers[_errorHandler.RetryCountHeader], Is.EqualTo("3"));
        Assert.That(delayedMessage.DeliverAt, Is.EqualTo(expectedRetryTime));
    }

    [Test]
    public async Task Retrier_ExceedsMaxDelayedRetries_SendsToDeadLetter()
    {
        // Arrange
        _incomingMessage.Headers[_errorHandler.RetryCountHeader] =
            _errorHandler.DelayedRetryCount.ToString();

        var testException = new Exception("Final error");

        Task Next() => throw testException;

        // Act
        await _errorHandler.Retrier(_transaction, Next);

        // Assert
        Assert.That(_transaction.OutgoingMessages.Count, Is.EqualTo(1));

        var deadLetterMessage = _transaction.OutgoingMessages[0];
        Assert.That(deadLetterMessage.Endpoint, Is.EqualTo(TestTransport.DefaultDeadLetterEndpoint));
        Assert.That(deadLetterMessage.Headers[_errorHandler.ErrorMessageHeader], Is.EqualTo("Final error"));
        Assert.That(deadLetterMessage.Headers[_errorHandler.ErrorStackTraceHeader], Is.Not.Null);
    }

    [Test]
    public async Task Retrier_ClearsOutgoingMessagesOnEachRetry()
    {
        // Arrange
        var callCount = 0;

        Task Next()
        {
            callCount++;
            _transaction.Add(new ErrorHandlerTestMessage());
            throw new Exception("Test error");
        }

        // Act
        await _errorHandler.Retrier(_transaction, Next);

        // Assert
        Assert.That(callCount, Is.EqualTo(_errorHandler.ImmediateRetryCount));
        // Should only have the delayed retry message, not the messages added in next()
        Assert.That(_transaction.OutgoingMessages.Count, Is.EqualTo(1));
        Assert.That(_transaction.OutgoingMessages[0].Headers.ContainsKey(_errorHandler.RetryCountHeader), Is.True);
    }

    [Test]
    public async Task Retrier_WithZeroImmediateRetries_SchedulesDelayedRetryImmediately()
    {
        // Arrange
        _errorHandler = new TestableErrorHandler { ImmediateRetryCount = 0 };
        var expectedRetryTime = DateTime.UtcNow.AddMinutes(5);
        _errorHandler.SetNextRetryTime(expectedRetryTime);

        var callCount = 0;

        Task Next()
        {
            callCount++;
            throw new Exception("Test error");
        }

        // Act
        await _errorHandler.Retrier(_transaction, Next);

        // Assert
        Assert.That(callCount, Is.EqualTo(1)); // Should enforce minimum of 1
        Assert.That(_transaction.OutgoingMessages.Count, Is.EqualTo(1));
        Assert.That(_transaction.OutgoingMessages[0].Headers[_errorHandler.RetryCountHeader], Is.EqualTo("1"));
    }

    [Test]
    public async Task Retrier_WithZeroDelayedRetries_StillGetsMinimumOfOne()
    {
        // Arrange
        _errorHandler = new TestableErrorHandler { DelayedRetryCount = 0 };
        var expectedRetryTime = DateTime.UtcNow.AddMinutes(5);
        _errorHandler.SetNextRetryTime(expectedRetryTime);

        Task Next() => throw new Exception("Test error");

        // Act
        await _errorHandler.Retrier(_transaction, Next);

        // Assert
        // Even with DelayedRetryCount = 0, Math.Max(1, DelayedRetryCount) makes it 1
        Assert.That(_transaction.OutgoingMessages.Count, Is.EqualTo(1));
        Assert.That(_transaction.OutgoingMessages[0].Headers[_errorHandler.RetryCountHeader], Is.EqualTo("1"));
        Assert.That(_transaction.OutgoingMessages[0].DeliverAt, Is.EqualTo(expectedRetryTime));
    }

    [Test]
    public void GetNextRetryTime_DefaultImplementation_UsesDelayCorrectly()
    {
        // Arrange
        var handler = new ErrorHandler { DelayIncrement = 60 };
        var beforeTime = DateTime.UtcNow;

        // Act
        var result1 = handler.GetNextRetryTime(1);
        var result2 = handler.GetNextRetryTime(2);
        var result3 = handler.GetNextRetryTime(3);

        var afterTime = DateTime.UtcNow;

        // Assert
        Assert.That(result1, Is.GreaterThanOrEqualTo(beforeTime.AddSeconds(60)));
        Assert.That(result1, Is.LessThanOrEqualTo(afterTime.AddSeconds(60)));

        Assert.That(result2, Is.GreaterThanOrEqualTo(beforeTime.AddSeconds(120)));
        Assert.That(result2, Is.LessThanOrEqualTo(afterTime.AddSeconds(120)));

        Assert.That(result3, Is.GreaterThanOrEqualTo(beforeTime.AddSeconds(180)));
        Assert.That(result3, Is.LessThanOrEqualTo(afterTime.AddSeconds(180)));
    }

    [Test]
    public async Task Retrier_SucceedsAfterSomeImmediateRetries_StopsRetrying()
    {
        // Arrange
        var callCount = 0;

        Task Next()
        {
            callCount++;
            if (callCount < 3) // Fail first 2 attempts
            {
                throw new Exception("Test error");
            }

            return Task.CompletedTask;
        }

        // Act
        await _errorHandler.Retrier(_transaction, Next);

        // Assert
        Assert.That(callCount, Is.EqualTo(3));
        Assert.That(_transaction.OutgoingMessages.Count, Is.EqualTo(0));
    }

    [Test]
    public async Task Retrier_InvalidRetryCountHeader_TreatsAsZero()
    {
        // Arrange
        _incomingMessage.Headers[_errorHandler.RetryCountHeader] = "invalid";
        var expectedRetryTime = DateTime.UtcNow.AddMinutes(5);
        _errorHandler.SetNextRetryTime(expectedRetryTime);

        Task Next() => throw new Exception("Test error");

        // Act
        await _errorHandler.Retrier(_transaction, Next);

        // Assert
        Assert.That(_transaction.OutgoingMessages.Count, Is.EqualTo(1));
        var delayedMessage = _transaction.OutgoingMessages[0];
        Assert.That(delayedMessage.Headers[_errorHandler.RetryCountHeader], Is.EqualTo("1"));
    }

    [Test]
    public async Task Retrier_ExceptionStackTrace_IsStoredInHeader()
    {
        // Arrange
        _incomingMessage.Headers[_errorHandler.RetryCountHeader] =
            _errorHandler.DelayedRetryCount.ToString();

        var exceptionWithStackTrace = new Exception("Error with stack trace");

        Task Next() => throw exceptionWithStackTrace;

        // Act
        await _errorHandler.Retrier(_transaction, Next);

        // Assert
        Assert.That(_transaction.OutgoingMessages.Count, Is.EqualTo(1));
        var deadLetterMessage = _transaction.OutgoingMessages[0];
        Assert.That(deadLetterMessage.Headers[_errorHandler.ErrorStackTraceHeader], Is.Not.Null);
        Assert.That(deadLetterMessage.Headers[_errorHandler.ErrorStackTraceHeader], Is.Not.Empty);
    }

    [Test]
    public async Task Retrier_ExceptionWithNullStackTrace_UsesEmptyString()
    {
        // Arrange
        _incomingMessage.Headers[_errorHandler.RetryCountHeader] =
            _errorHandler.DelayedRetryCount.ToString();

        // Create an exception with null StackTrace using custom exception
        var exceptionWithoutStackTrace = new ExceptionWithNullStackTrace("Error without stack trace");

        Task Next() => throw exceptionWithoutStackTrace;

        // Act
        await _errorHandler.Retrier(_transaction, Next);

        // Assert
        Assert.That(_transaction.OutgoingMessages.Count, Is.EqualTo(1));
        var deadLetterMessage = _transaction.OutgoingMessages[0];
        Assert.That(deadLetterMessage.Headers[_errorHandler.ErrorStackTraceHeader], Is.EqualTo(string.Empty));
    }

    // Helper class to override GetNextRetryTime for testing
    private class TestableErrorHandler : ErrorHandler
    {
        private DateTime? _nextRetryTime;

        public void SetNextRetryTime(DateTime retryTime)
        {
            _nextRetryTime = retryTime;
        }

        public override DateTime GetNextRetryTime(int attempt)
        {
            return _nextRetryTime ?? base.GetNextRetryTime(attempt);
        }
    }
}

File: src/kotlin/src/test/kotlin/polybus/transport/transactions/messages/handlers/error/ErrorHandlerTests.kt
package polybus.transport.transactions.messages.handlers.error

import kotlinx.coroutines.runBlocking
import org.junit.jupiter.api.BeforeEach
import org.junit.jupiter.api.Assertions.assertEquals
import org.junit.jupiter.api.Assertions.assertFalse
import org.junit.jupiter.api.Assertions.assertNotNull
import org.junit.jupiter.api.Assertions.assertTrue
import org.junit.jupiter.api.Test
import polybus.transport.transactions.IncomingTransaction
import polybus.transport.transactions.messages.IncomingMessage
import polybus.transport.transactions.messages.handlers.error.ErrorHandler
import java.time.Instant
import java.time.temporal.ChronoUnit

class ErrorHandlerTests {
    private lateinit var testBus: TestBus
    private lateinit var transaction: IncomingTransaction
    private lateinit var errorHandler: TestableErrorHandler

    @BeforeEach
    fun setUp() {
        testBus = TestBus("TestBus")
        testBus.messages.add(ErrorHandlerTestMessage::class.java)
        val incomingMessage = IncomingMessage(
            bus = testBus,
            body = "{}",
            messageInfo = testBus.messages.getMessageInfo(ErrorHandlerTestMessage::class.java)
        )
        transaction = IncomingTransaction(testBus, incomingMessage)
        errorHandler = TestableErrorHandler()
    }

    @Test
    fun retrierSucceedsOnFirstAttemptDoesNotRetry() = runBlocking {
        // Arrange
        var nextCalled = false

        // Act
        errorHandler.retrier(transaction) {
            nextCalled = true
        }

        // Assert
        assertTrue(nextCalled)
        assertEquals(0, transaction.outgoingMessages.size)
    }

    @Test
    fun retrierFailsOnceRetriesImmediately() = runBlocking {
        // Arrange
        var callCount = 0

        // Act
        errorHandler.retrier(transaction) {
            callCount++
            if (callCount == 1) {
                throw RuntimeException("Test error")
            }
        }

        // Assert
        assertEquals(2, callCount)
        assertEquals(0, transaction.outgoingMessages.size)
    }

    @Test
    fun retrierFailsAllImmediateRetriesSchedulesDelayedRetry() = runBlocking {
        // Arrange
        val expectedRetryTime = Instant.now().plusSeconds(300)
        errorHandler.setNextRetryTime(expectedRetryTime)
        var callCount = 0

        // Act
        errorHandler.retrier(transaction) {
            callCount++
            throw RuntimeException("Test error")
        }

        // Assert
        assertEquals(errorHandler.immediateRetryCount, callCount)
        assertEquals(1, transaction.outgoingMessages.size)

        val delayedMessage = transaction.outgoingMessages.first()
        assertEquals(expectedRetryTime, delayedMessage.deliverAt)
        assertEquals("1", delayedMessage.headers[errorHandler.retryCountHeader])
        assertEquals("TestBus", delayedMessage.endpoint)
    }

    @Test
    fun retrierWithExistingRetryCountIncrementsCorrectly() = runBlocking {
        // Arrange
        transaction.incomingMessage.headers[errorHandler.retryCountHeader] = "2"
        val expectedRetryTime = Instant.now().plusSeconds(600)
        errorHandler.setNextRetryTime(expectedRetryTime)

        // Act
        errorHandler.retrier(transaction) {
            throw RuntimeException("Test error")
        }

        // Assert
        assertEquals(1, transaction.outgoingMessages.size)

        val delayedMessage = transaction.outgoingMessages.first()
        assertEquals("3", delayedMessage.headers[errorHandler.retryCountHeader])
        assertEquals(expectedRetryTime, delayedMessage.deliverAt)
    }

    @Test
    fun retrierExceedsMaxDelayedRetriesSendsToDeadLetter() = runBlocking {
        // Arrange
        transaction.incomingMessage.headers[errorHandler.retryCountHeader] =
            errorHandler.delayedRetryCount.toString()
        val testException = RuntimeException("Final error")

        // Act
        errorHandler.retrier(transaction) {
            throw testException
        }

        // Assert
        assertEquals(1, transaction.outgoingMessages.size)

        val deadLetterMessage = transaction.outgoingMessages.first()
        assertEquals(TestTransport.DEFAULT_DEAD_LETTER_ENDPOINT, deadLetterMessage.endpoint)
        assertEquals("Final error", deadLetterMessage.headers[errorHandler.errorMessageHeader])
        assertNotNull(deadLetterMessage.headers[errorHandler.errorStackTraceHeader])
    }

    @Test
    fun retrierClearsOutgoingMessagesOnEachRetry() = runBlocking {
        // Arrange
        var callCount = 0

        // Act
        errorHandler.retrier(transaction) {
            callCount++
            transaction.add(ErrorHandlerTestMessage())
            throw RuntimeException("Test error")
        }

        // Assert
        assertEquals(errorHandler.immediateRetryCount, callCount)
        // Should only have the delayed retry message, not the messages added in next()
        assertEquals(1, transaction.outgoingMessages.size)
        assertTrue(
            transaction.outgoingMessages.first().headers.containsKey(errorHandler.retryCountHeader)
        )
    }

    @Test
    fun retrierWithZeroImmediateRetriesSchedulesDelayedRetryImmediately() = runBlocking {
        // Arrange
        errorHandler.immediateRetryCount = 0
        val expectedRetryTime = Instant.now().plusSeconds(300)
        errorHandler.setNextRetryTime(expectedRetryTime)
        var callCount = 0

        // Act
        errorHandler.retrier(transaction) {
            callCount++
            throw RuntimeException("Test error")
        }

        // Assert
        assertEquals(1, callCount) // Should enforce minimum of 1
        assertEquals(1, transaction.outgoingMessages.size)
        assertEquals("1", transaction.outgoingMessages.first().headers[errorHandler.retryCountHeader])
    }

    @Test
    fun retrierWithZeroDelayedRetriesStillGetsMinimumOfOne() = runBlocking {
        // Arrange
        errorHandler.delayedRetryCount = 0
        val expectedRetryTime = Instant.now().plusSeconds(300)
        errorHandler.setNextRetryTime(expectedRetryTime)

        // Act
        errorHandler.retrier(transaction) {
            throw RuntimeException("Test error")
        }

        // Assert
        assertEquals(1, transaction.outgoingMessages.size)
        assertEquals("1", transaction.outgoingMessages.first().headers[errorHandler.retryCountHeader])
        assertEquals(expectedRetryTime, transaction.outgoingMessages.first().deliverAt)
    }

    @Test
    fun getNextRetryTimeDefaultImplementationUsesDelayCorrectly() {
        // Arrange
        val handler = ErrorHandler().apply { delayIncrement = 60 }
        val beforeTime = Instant.now()

        // Act
        val result1 = handler.getNextRetryTime(1)
        val result2 = handler.getNextRetryTime(2)
        val result3 = handler.getNextRetryTime(3)

        val afterTime = Instant.now()

        // Assert
        assertTrue(result1 >= beforeTime.plus(60, ChronoUnit.SECONDS))
        assertTrue(result1 <= afterTime.plus(60, ChronoUnit.SECONDS))

        assertTrue(result2 >= beforeTime.plus(120, ChronoUnit.SECONDS))
        assertTrue(result2 <= afterTime.plus(120, ChronoUnit.SECONDS))

        assertTrue(result3 >= beforeTime.plus(180, ChronoUnit.SECONDS))
        assertTrue(result3 <= afterTime.plus(180, ChronoUnit.SECONDS))
    }

    @Test
    fun retrierSucceedsAfterSomeImmediateRetriesStopsRetrying() = runBlocking {
        // Arrange
        var callCount = 0

        // Act
        errorHandler.retrier(transaction) {
            callCount++
            if (callCount < 3) {
                throw RuntimeException("Test error")
            }
        }

        // Assert
        assertEquals(3, callCount)
        assertEquals(0, transaction.outgoingMessages.size)
    }

    @Test
    fun retrierInvalidRetryCountHeaderTreatsAsZero() = runBlocking {
        // Arrange
        transaction.incomingMessage.headers[errorHandler.retryCountHeader] = "invalid"
        val expectedRetryTime = Instant.now().plusSeconds(300)
        errorHandler.setNextRetryTime(expectedRetryTime)

        // Act
        errorHandler.retrier(transaction) {
            throw RuntimeException("Test error")
        }

        // Assert
        assertEquals(1, transaction.outgoingMessages.size)
        val delayedMessage = transaction.outgoingMessages.first()
        assertEquals("1", delayedMessage.headers[errorHandler.retryCountHeader])
    }

    @Test
    fun retrierExceptionStackTraceIsStoredInHeader() = runBlocking {
        // Arrange
        transaction.incomingMessage.headers[errorHandler.retryCountHeader] =
            errorHandler.delayedRetryCount.toString()

        val exceptionWithStackTrace = RuntimeException("Error with stack trace")

        // Act
        errorHandler.retrier(transaction) {
            throw exceptionWithStackTrace
        }

        // Assert
        assertEquals(1, transaction.outgoingMessages.size)
        val deadLetterMessage = transaction.outgoingMessages.first()
        val storedStackTrace = deadLetterMessage.headers[errorHandler.errorStackTraceHeader]
        assertNotNull(storedStackTrace)
        assertFalse(storedStackTrace.isNullOrEmpty())
    }

    @Test
    fun retrierExceptionWithNullStackTraceUsesEmptyString() = runBlocking {
        // Arrange
        transaction.incomingMessage.headers[errorHandler.retryCountHeader] =
            errorHandler.delayedRetryCount.toString()

        // Create an exception with null StackTrace using custom exception
        val exceptionWithoutStackTrace = RuntimeException("Error without stack trace")
        exceptionWithoutStackTrace.stackTrace = emptyArray()

        // Act
        errorHandler.retrier(transaction) {
            throw exceptionWithoutStackTrace
        }

        // Assert
        assertEquals(1, transaction.outgoingMessages.size)
        val deadLetterMessage = transaction.outgoingMessages.first()
        assertEquals("", deadLetterMessage.headers[errorHandler.errorStackTraceHeader])
    }

    private class TestableErrorHandler : ErrorHandler() {
        private var _nextRetryTime: Instant? = null

        fun setNextRetryTime(retryTime: Instant) {
            _nextRetryTime = retryTime
        }

        override fun getNextRetryTime(attempt: Int): Instant {
            return _nextRetryTime ?: super.getNextRetryTime(attempt)
        }
    }
}
